src/framework/ui/uimanager.cpp

```diff
@@ -407,26 +431,89 @@ std::string UIManager::getStyleClass(const std::string_view styleName)
     return "";
 }
 
+OTMLNodePtr UIManager::findMainWidgetNode(const OTMLDocumentPtr& doc)
+{
+    OTMLNodePtr mainNode = nullptr;
+    for (const auto& node : doc->children()) {
+        std::string tag = node->tag();
+
+        if (tag.find('<') == std::string::npos) {
+            if (mainNode)
+                throw Exception("cannot have multiple main widgets in otui files");
+            mainNode = node;
+        }
+    }
+    return mainNode;
+}
+
+OTMLNodePtr UIManager::loadDeviceUI(const std::string& file, Platform::OperatingSystem os)
+{
+    auto rawName = file.substr(0, file.find("."));
+    auto osName = g_platform.getOsShortName(os);
+
+    const auto& doc = OTMLDocument::parse(g_resources.guessFilePath(rawName + "." + osName, "otui"));
+    if (doc) {
+        g_logger.info(stdext::format("found os style '%s' for '%s'", osName, rawName));
+        importStyleFromOTML(doc);
+        return findMainWidgetNode(doc);
+    }
+    return nullptr;
+}
+
+OTMLNodePtr UIManager::loadDeviceUI(const std::string& file, Platform::DeviceType deviceType)
+{
+    auto rawName = file.substr(0, file.find("."));
+    auto deviceName = g_platform.getDeviceShortName(deviceType);
+
+    const auto& doc = OTMLDocument::parse(g_resources.guessFilePath(rawName + "." + deviceName, "otui"));
+    if (doc) {
+        g_logger.info(stdext::format("found device style '%s' for '%s'", deviceName, rawName));
+        importStyleFromOTML(doc);
+        return findMainWidgetNode(doc);
+    }
+    return nullptr;
+}
+
 UIWidgetPtr UIManager::loadUI(const std::string& file, const UIWidgetPtr& parent)
 {
     try {
+        OTMLNodePtr widgetNode = nullptr;
         const auto& doc = OTMLDocument::parse(g_resources.guessFilePath(file, "otui"));
-        UIWidgetPtr widget;
+
         for (const auto& node : doc->children()) {
             std::string tag = node->tag();
 
             // import styles in these files too
             if (tag.find('<') != std::string::npos)
                 importStyleFromOTML(node);
             else {
-                if (widget)
+                if (widgetNode)
                     throw Exception("cannot have multiple main widgets in otui files");
-                widget = createWidgetFromOTML(node, parent);
+                widgetNode = node;
             }
         }
 
-        return widget;
-    } catch (stdext::exception& e) {
+        // load device styles and widget
+        auto device = g_platform.getDevice();
+        try {
+            const auto& deviceWidgetNode = loadDeviceUI(file, device.type);
+            if (deviceWidgetNode)
+                widgetNode = deviceWidgetNode;
+        } catch (stdext::exception&) {}
```

Comment:
Wouldn't it be interesting to send something in the exception? Maybe a log reporting the error.

Reply:
We could do a `fine` log output here, but its expected that a lot of style loading will come back with this exception. So to me it doesn't really give us much information even for debug purposes, you're better off going into the debugger.

Reply:
However if there is an exception we don't expect we dont want to swallow it, so I will add a fine log output. Thanks for the suggestion!